source/agora/test/FutureEnrollment.d

```diff
@@ -0,0 +1,114 @@
+///     enrollment.
+/// Expectation: The slow validator accepts the enrollment even though
+///     there is not the frozen UTXO in the UTXO set because the
+///     transaction that contains the UTXO exists in the transaction pool.
+unittest
+{
+    static class CustomValidator : TestValidatorNode
+    {
+        mixin ForwardCtor!();
+
+        ///
+        protected override void catchupTask () nothrow
+        {
+            // Do not catch up to the network
+            return;
+        }
+    }
+
+    static class CustomAPIManager : TestAPIManager
+    {
+        mixin ForwardCtor!();
+
+        /// set base class
+        public override void createNewNode (Config conf, string file, int line)
+        {
+            if (this.nodes.length == GenesisValidators - 1)
+                this.addNewNode!CustomValidator(conf);
+            else
+                super.createNewNode(conf, file, line);
+        }
+    }
+
+    TestConf conf = {
+        outsider_validators : 1,
+        recurring_enrollment : false
+    };
+    auto all_validators = GenesisValidators + conf.outsider_validators;
+    auto network = makeTestNetwork!CustomAPIManager(conf);
+    network.start();
+    scope(exit) network.shutdown();
+    scope(failure) network.printLogs();
+    network.waitForDiscovery();
+
+    auto slow_node = GenesisValidators - 1;
+    auto outsider = network.nodes[GenesisValidators];
+
+    network.generateBlocks(Height(GenesisValidatorCycle - 2));
+
+    // make sure outsiders are up to date
+    network.expectHeight(iota(GenesisValidators, all_validators),
+        Height(GenesisValidatorCycle - 2));
+
+    // prepare frozen outputs for the outsider validator to enroll
+    const key = outsider.getPublicKey().key;
+    network.blocks[0].spendable().drop(1).takeExactly(1)
+        .map!(txb => txb
+            .split([key]).sign(OutputType.Freeze))
+            .each!(tx => network.nodes[slow_node].postTransaction(tx));
+
+    // the bad validator becomes unresponsive
+    network.clients[slow_node].filter!(API.postTransaction);
+
+    // Block 19 we add the frozen utxo for the outsider validator
+    auto target_nodes = iota(GenesisValidators - 1).array;
+    target_nodes ~= GenesisValidators;
+    network.generateBlocks(target_nodes, Height(GenesisValidatorCycle - 1));
+
+    // the bad validator is a block behind from the latest height
+    assert(network.clients[slow_node].getBlockHeight() == GenesisValidatorCycle - 2);
+ 
+    // enroll the outsider
+    auto enroll = network.clients[GenesisValidators].setRecurringEnrollment(true);
+    assert(enroll != Enrollment.init);
+
+    // re-enroll all the validators
+    iota(GenesisValidators).each!(i => network.enroll(iota(GenesisValidators), i));
+
+    // check that the bad validator has the enrollment of the outsider
+    auto gotten_enroll = network.clients[slow_node].getEnrollment(enroll.utxo_key);
+    assert(gotten_enroll == enroll);
+
+    network.clients[slow_node].clearFilter();
```

Comment:
There should be a check that a block is externalized with an enrollment matching the frozen UTXO.

Reply:
Right. Generating a block having the enrollment makes the test complete.